nettests/websites/web_connectivity.go

```diff
@@ -72,17 +41,16 @@ type WebConnectivity struct {
 
 // Run starts the test
 func (n WebConnectivity) Run(ctl *nettests.Controller) error {
-	nt := mk.NewNettest("WebConnectivity")
-	ctl.Init(nt)
-
 	urls, urlIDMap, err := lookupURLs(ctl)
 	if err != nil {
 		return err
 	}
 	ctl.SetInputIdxMap(urlIDMap)
-	nt.Options.Inputs = urls
-
-	return nt.Run()
+	experiment := web_connectivity.NewExperiment(
```

Comment:
I believe the logic of how this test now runs has significantly changed. It now does a GeoIP lookup for every URL and opens a report for each measurement.
This significantly increases the overhead of running a test and the progress in it is also not as clear.

We should see if we can make this change in such a way that it implements the same flow as before.

Reply:
What leads you to say so? I'm using MK v0.10.4 and I cannot see this behaviour. The geolookup and report management functionality are now delegated to the Go engine and MK is explicitly configured to basically just run the measurement. However, it may be that with an older version of MK some functionality is duplicated because I've added support for disabling more functionality only in v0.10.4.

Please, see below an excerpt of the verbose log where I dissect what happens. This seems to me a long but also data-centred way of understanding if we're doing in MK some bits of processing that we should not be doing now. (Spoiler alert: I found a couple of issues that need fixing but AFAICT we're not doing unnecessary geolookups or collector calls.) While there, I have also added comments on how we could further improve the performance.

First, ooni/probe-cli starts up and prints some information:

```
   • ooni version 3.0.0-beta.3
   • Reading default config file
   • Connecting to database sqlite3:///Users/sbasso/.ooni/db/main.sqlite3
   • running migrations       
   • performed 0 migrations   
   • Using collector:         
   • Using bouncer:           
```

Then, ooni/probe-engine performs the GeoIP lookup:

```
[engine] iplookup: using akamai
[engine] dns: resolving a248.e.akamai.net
[engine] dns: got 2 entries
[engine] - 92.123.101.210
[engine] - 84.53.177.17
[engine] connect: using tcp, 92.123.101.210:443
[engine] connect: connected to tcp, 92.123.101.210:443
[engine] tls: starting handshake
[engine] tls: handshake OK
[engine] - negotiated protocol: http/1.1
[engine] - version: TLSv1.2
[engine] http: connection to 92.123.101.210:443 ready; sending request
[engine] > GET / HTTP/1.1
[engine] > User-Agent: ooniprobe-desktop/3.0.0-beta.3
[engine] > Accept-Encoding: gzip
[engine] > Host: whatismyip.akamai.com
[engine] >
[engine] http: request sent; waiting for response
[engine] http: start receiving response
[engine] < HTTP/1.1 200 OK
[engine] < Expires: Sun, 16 Jun 2019 07:54:45 GMT
[engine] < Cache-Control: max-age=0, no-cache, no-store
[engine] < Pragma: no-cache
[engine] < Date: Sun, 16 Jun 2019 07:54:45 GMT
[engine] < Connection: keep-alive
[engine] < Content-Type: text/html
[engine] < Content-Length: 14
[engine] <
[engine] iplookup: IP: 109.116.253.87
[engine] mmdblookup: ASN: &{AutonomousSystemNumber:30722 AutonomousSystemOrganization:Vodafone Italia S.p.A.}
[engine] mmdblookup: Country: &{Country:{IsoCode:IT ...}
```

Then, ooni/probe-engine queries the bouncer for backends (if we want this to be more efficient, we should probably also have a single API call returning all the info):

```
[engine] jsonapi: method: GET
[engine] jsonapi: URL: https://bouncer.ooni.io/api/v1/collectors
[engine] dns: resolving bouncer.ooni.io
[engine] dns: got 1 entries
[engine] - 103.104.244.87
[engine] connect: using tcp, 103.104.244.87:443
[engine] connect: connected to tcp, 103.104.244.87:443
[engine] tls: starting handshake
[engine] tls: handshake OK
[engine] - negotiated protocol: http/1.1
[engine] - version: TLSv1.2
[engine] http: connection to 103.104.244.87:443 ready; sending request
[engine] > GET /api/v1/collectors HTTP/1.1
[engine] > Accept-Encoding: gzip
[engine] > User-Agent: ooniprobe-desktop/3.0.0-beta.3
[engine] > Host: bouncer.ooni.io
[engine] >
[engine] http: request sent; waiting for response
[engine] http: start receiving response
[engine] < HTTP/1.1 200 OK
[engine] < Date: Sun, 16 Jun 2019 07:54:46 GMT
[engine] < Content-Type: application/json
[engine] < Content-Length: 466
[engine] < Connection: keep-alive
[engine] < Keep-Alive: timeout=120
[engine] < Etag: "f78333f849c8eaa385ba2fc03fcb69c758585c6b"
[engine] < Server: nginx
[engine] <
[engine] jsonapi: response body: [
    {"type": "https", "address": "https://c.collector.ooni.io:443"}
, ...]
[engine] jsonapi: method: GET
[engine] jsonapi: URL: https://bouncer.ooni.io/api/v1/test-helpers
[engine] http: connection to 103.104.244.87:443 ready; sending request
```

However, note how using a global HTTP client is nice because here we're clearly reusing the connection used for calling the previous API (so we already save some round-trips).

```
[engine] > GET /api/v1/test-helpers HTTP/1.1
[engine] > Host: bouncer.ooni.io
[engine] > Accept-Encoding: gzip
[engine] > User-Agent: ooniprobe-desktop/3.0.0-beta.3
[engine] >
[engine] http: request sent; waiting for response
[engine] http: start receiving response
[engine] < HTTP/1.1 200 OK
[engine] < Server: nginx
[engine] < Date: Sun, 16 Jun 2019 07:54:47 GMT
[engine] < Content-Type: application/json
[engine] < Content-Length: 1126
[engine] < Connection: keep-alive
[engine] < Keep-Alive: timeout=120
[engine] < Etag: "953c84ce7845161ff0c99fcb881e4efdebfa6b5e"
[engine] <
[engine] jsonapi: response body: {
    "traceroute": [{"type": "legacy", "address": "37.218.247.110"},  ...
]}
```

Then, ooni/probe-cli prepares for running `websites`:

```
   • Running websites         
   • Running test group Websites
   • Creating result {
    0 websites 2019-06-16 07:54:47.128591 +0000 UTC
    16 0 false false 0 0
    $HOME/.ooni/msmts/websites-2019-06-16T075447.128591Z}
   • Running test websites.WebConnectivity
```

As part of this, ooni/probe-engine queries the test-lists API (if this API was served by the same domain as above, this would mean continuing to reuse the same connection):

```
[engine] jsonapi: method: GET
[engine] jsonapi: URL: https://events.proteus.ooni.io/api/v1/urls?probe_cc=IT
```

Incidentally, this should be `events.ooni.io` (I probably need to `go get -v`):

```
[engine] dns: resolving events.proteus.ooni.io
[engine] dns: got 1 entries
[engine] - 37.218.242.63
[engine] connect: using tcp, 37.218.242.63:443
[engine] connect: connected to tcp, 37.218.242.63:443
[engine] tls: starting handshake
[engine] tls: handshake OK
[engine] - negotiated protocol: http/1.1
[engine] - version: TLSv1.2
[engine] http: connection to 37.218.242.63:443 ready; sending request
[engine] > GET /api/v1/urls?probe_cc=IT HTTP/1.1
[engine] > User-Agent: ooniprobe-desktop/3.0.0-beta.3
[engine] > Host: events.proteus.ooni.io
[engine] > Accept-Encoding: gzip
[engine] >
[engine] http: request sent; waiting for response
[engine] http: start receiving response
[engine] < HTTP/1.1 200 OK
[engine] < Connection: keep-alive
[engine] < Server: nginx/1.6.2
[engine] < Date: Sun, 16 Jun 2019 07:54:47 GMT
[engine] < Content-Type: application/json; charset=utf-8
[engine] <
[engine] jsonapi: response body: {"metadata":{
    "count":-1,"current_page":-1,"limit":100,"next_url":"","pages":-1
},"results":[{
    "category_code":"COMT",
    "url":"https://www.iconnecthere.com/","country_code":"XX"},
...}]}
```

Then ooni/probe-cli processes all the URLs:

```
   • Going over URL 0         
   • returning url 198        
   • Mapped URL https://www.iconnecthere.com/ to idx 0 and urlID 198
   • Going over URL 1         
   • returning url 199        
[...]
   • Going over URL 99        
   • returning url 283        
   • Mapped URL http://emchurch.org/ to idx 99 and urlID 283
```

Then, the nettest proper starts (note that `status.queue` and `status.start` are not MK events but I've kept the output consistent with what we used to emit):

```
   • status.queued            
   • status.started           
   • OutputPath: /.../ooni283065313/msmt-websites.WebConnectivity-2019-06-16T075447.134412Z.jsonl
```

Then, the ooni/probe-engine is again invoked to open the report (same consideration as above with respect to possibly reusing connections if reusing domain names):

```
[engine] experiment: unsupported collector type: onion
[engine] jsonapi: request body: {
  "probe_asn":"AS30722",
  "probe_cc":"IT",
  "software_name":"ooniprobe-desktop",
  "software_version":"3.0.0-beta.3",
  "test_name":"web_connectivity",
  "test_version":"0.0.1"
}
[engine] jsonapi: method: POST
[engine] jsonapi: URL: https://c.collector.ooni.io:443/report
[engine] dns: resolving c.collector.ooni.io
[engine] dns: got 1 entries
[engine] - 37.218.242.210
[engine] connect: using tcp, 37.218.242.210:443
[engine] connect: connected to tcp, 37.218.242.210:443
[engine] tls: starting handshake
[engine] tls: handshake OK
[engine] - negotiated protocol: http/1.1
[engine] - version: TLSv1.2
[engine] http: connection to 37.218.242.210:443 ready; sending request
[engine] > POST /report HTTP/1.1
[engine] > Content-Length: 163
[engine] > Content-Type: application/json
[engine] > Host: c.collector.ooni.io:443
[engine] > Accept-Encoding: gzip
[engine] > User-Agent: ooniprobe-desktop/3.0.0-beta.3
[engine] >
[engine] http: request sent; waiting for response
[engine] http: start receiving response
[engine] < HTTP/1.1 200 OK
[engine] < Server: nginx/1.6.2
[engine] < Date: Sun, 16 Jun 2019 07:54:47 GMT
[engine] < Content-Type: application/json
[engine] < Content-Length: 159
[engine] < Connection: keep-alive
[engine] < Keep-Alive: timeout=120
[engine] <
[engine] jsonapi: response body: {
  "backend_version": "1.3.5",
  "supported_formats": ["yaml", "json"],
  "report_id": "..."
}
```

At this point (and excuse me for this long message, but I wanted to be sure what I thought and the code behaviour were consistent), we finally invoke MK. I will dissect the settings that we're passing to it, which is functional to explaining what is happening:

```
   • status.report_create     
   • status.measurement_start 
[engine] measurementkit: settings: {
```

As you can see, we're invoking MK's Web Connectivity for a single input. (The log level is DEBUG because I wanted to show you more logs here.)

```
  "inputs":["https://www.iconnecthere.com/"],
  "log_level":"DEBUG",
  "name":"WebConnectivity",
```

Note how we're passing to MK the backend that we discovered using ooni/probe-engine and how we're also passing the paths to the assets downloaded by the engine:

```
  "options":{
    "backend":"https://c.web-connectivity.th.ooni.io:443",
    "net/ca_bundle_path":"$HOME/.ooni/assets/ca-bundle.pem",
    "geoip_asn_path":"$HOME/.ooni/assets/asn.mmdb",
```

Note how we disable bouncer, collector, saving report to disk, and GeoIP lookup:

```
    "no_bouncer":true,
    "no_collector":true,
    "no_file_report":true,
    "no_geoip":true,
```

Note how we're passing downstream all the information that we have:

```
    "probe_asn":"AS30722",
    "probe_cc":"IT",
    "probe_ip":"109.116.253.87",
    "probe_network_name":"Vodafone Italia S.p.A.",
```

That we're passing `save_real_probe_ip` equal to `true` looks like a bug to me, in the sense that this disables the IP scrubbing performed by MK.

```
    "save_real_probe_ip":true,
    "software_name":"ooniprobe-desktop",
    "software_version":"3.0.0-beta.3"
}}
```

Then MK will perform the resolver lookup. This is actually performed for each measurement and I believe this is a bug and we should do it only once in the engine:

```
[engine] getaddrinfo('whoami.akamai.net') => error: code=0, reason=''
[engine] ip address of resolver is 173.194.169.102
```

Then there is the measurement proper. MK runs three measurement in parallel, but since we're passing it just one entry, notice how two threads out of three immediately reach the end of the input and basically terminate. (We should discuss whether we want parallelism in ooni/probe-cli or not; for now it is not implemented because it would have complicated too much the review of this diff, but clearly a subsequent PR can improve the situation, if we want.)

```
   • OnProgress: 0.100000 - starting the test
   0.10% starting the test        
[engine] net_test: running next measurement
   • OnProgress: 0.100000 - Processing input: https://www.iconnecthere.com/
   0.10% Processing input: https://www.iconnecthere.com/
[engine] net_test: creating entry
[engine] net_test: calling setup
[engine] net_test: running with input https://www.iconnecthere.com/
[engine] web_connectivity: starting dns_query for www.iconnecthere.com
[engine] net_test: running next measurement
[engine] net_test: reached end of input
[engine] net_test: running next measurement
[engine] net_test: reached end of input
```

See how Web Connectivity is performed for a single URL (with some snipping):

```
[engine] getaddrinfo('www.iconnecthere.com') => error: code=0, reason=''
[engine] dns_test: got response!
[engine] dns_test: callbacking
[engine] web_connectivity: starting tcp_connect
[engine] resolve_hostname: 163.44.187.153
[engine] resolve_hostname: is valid ipv4
[engine] connect_base 163.44.187.153:443
[engine] connect() in progress...
[engine] dns_test: callback called
[engine] connect time: 0.315201
[engine] web_connectivity: success to connect to 163.44.187.153:443
[engine] web_connectivity: starting http_request to https://www.iconnecthere.com/
[engine] Requesting url https://www.iconnecthere.com/
[engine] resolve_hostname: www.iconnecthere.com
[engine] resolve_hostname: ipv4...
[engine] getaddrinfo('www.iconnecthere.com') => error: code=0, reason=''
[engine] resolve_hostname: ipv4... done
[engine] resolve_hostname: ipv6...
[engine] getaddrinfo('www.iconnecthere.com') => error: code=0, reason=''
[engine] resolve_hostname: ipv6... done
[engine] connect_base 163.44.187.153:443
[engine] connect() in progress...
[engine] connect time: 0.271237
[engine] ssl: creating SSL_CTX with bundle: '/Users/sbasso/.ooni/assets/ca-bundle.pem'
[engine] Re-enabling SSLv2 and SSLv3
[engine] SSL version: TLSv1.2
[engine] ssl: handshake...
[engine] ssl: handshake... complete
[engine] Allowing dirty SSL shutdown
[engine] > GET / HTTP/1.1
[engine] > User-Agent: Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36
[engine] > Accept-Language: en-US;q=0.8,en;q=0.5
[engine] > Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
[engine] > Host: www.iconnecthere.com
[engine] >
[engine] http: started reading response
[engine] http: passing read data to parser
[engine] < HTTP/1.1 200 OK
[engine] < Server: nginx
[engine] < Date: Sun, 16 Jun 2019 07:54:49 GMT
[engine] < Content-Type: text/html; charset=UTF-8
[engine] < Transfer-Encoding: chunked
[engine] < Connection: keep-alive
[engine] < Link: <https://www.iconnecthere.com/wp-json/>; rel="https://api.w.org/", <https://wp.me/P9whHu-Bz>; rel=shortlink
[engine] < X-XSS-Protection: 1; mode=block
[engine] < X-Content-Type-Options: nosniff
[engine] < X-Nginx-Cache: HIT
[engine] <
[engine] http: continue reading for the response
[engine] http: passing read data to parser
...
[engine] http: received error 0 on connection
[engine] web_connectivity: doing control request
[engine] Using backend https://c.web-connectivity.th.ooni.io:443
[engine] resolve_hostname: c.web-connectivity.th.ooni.io
[engine] resolve_hostname: ipv4...
[engine] getaddrinfo('c.web-connectivity.th.ooni.io') => error: code=0, reason=''
[engine] resolve_hostname: ipv4... done
[engine] resolve_hostname: ipv6...
[engine] getaddrinfo('c.web-connectivity.th.ooni.io') => error: code=0, reason=''
[engine] resolve_hostname: ipv6... done
[engine] connect_base 37.218.241.164:443
[engine] connect() in progress...
[engine] connect time: 0.564001
[engine] SSL version: TLSv1.2
[engine] ssl: handshake...
[engine] ssl: handshake... complete
[engine] > POST / HTTP/1.1
[engine] > Content-Type: application/json
[engine] > Host: c.web-connectivity.th.ooni.io
[engine] > Content-Length: 350
[engine] >
[engine] http: started reading response
[engine] http: passing read data to parser
[engine] < HTTP/1.1 200 OK
[engine] < Date: Sun, 16 Jun 2019 07:54:52 GMT
[engine] < Content-Length: 559
[engine] < Content-Type: application/json
[engine] < Server: cyclone/1.1
[engine] <
[engine] http: received error 0 on connection
[engine] web_connectivity: comparing control with experiment
[engine] web_connectivity: comparing http_requests
[engine] web_connectivity: body proportion 1.000000
[engine] web_connectivity: comparing status codes
[engine] web_connectivity: checking headers
[engine] web_connectivity: checking HTML titles
[engine] web_connectivity: comparing dns_queries
[engine] web_connectivity: comparing tcp_connect
[engine] web_connectivity: exp,ctrl http failure determination
[engine] web_connectivity: checking if we got expected web page
[engine] web_connectivity: determining blocking reason
[engine] web_connectivity: no blocking detected
```

See how we're tearing down the nettest and leaving MK:

```
[engine] net_test: tearing down
[engine] net_test: written entry
[engine] net_test: running next measurement
[engine] net_test: reached end of input
   • OnProgress: 0.950000 - ending the test
   0.95% ending the test          
   • OnProgress: 1.000000 - test complete
   1.00% test complete            
```

Now the control's back to ooni/probe-cli and ooni/probe-engine. What we're doing now is submitting the measurement to the collector.

```
[engine] jsonapi: request body: {
    "content": {
        "annotations": {
            "engine_name": "libmeasurement_kit",
            "engine_version": "0.10.4",
            "engine_version_full": "v0.10.4",
            "platform": "macos"
        },
        "data_format_version": "0.2.0",
        "id": "8139bf0d-53b6-400b-8f3a-39a020cfd046",
        "input": "https://www.iconnecthere.com/",
        "measurement_start_time": "2019-06-16 07:54:47",
        "probe_asn": "AS30722",
        "probe_cc": "IT",
        "probe_ip": "127.0.0.1",
        "report_id": "",
        "software_name": "ooniprobe-desktop",
        "software_version": "3.0.0-beta.3",
        "test_helpers": {
            "backend": {
                "address": "https://c.web-connectivity.th.ooni.io:443",
                "type": "https"
            }
        },
        "test_keys": {
            "accessible": true,
            "agent": "redirect",
            "blocking": false,
            "body_length_match": true,
            "body_proportion": 1,
            "client_resolver": "173.194.169.102",
            "control": {
                "dns": {
                    "addrs": [
                        "163.44.187.153"
                    ],
                    "failure": null
                },
                "http_request": {
                    "body_length": 554651,
                    "failure": null,
                    "headers": {
                        "Content-Type": "text/html; charset=UTF-8",
                        "Date": "Sun, 16 Jun 2019 07:54:53 GMT",
                        "Link": "<https://www.iconnecthere.com/wp-json/>; rel=\"https://api.w.org/\", <https://wp.me/P9whHu-Bz>; rel=shortlink",
                        "Server": "nginx",
                        "X-Content-Type-Options": "nosniff",
                        "X-Nginx-Cache": "HIT",
                        "X-XSS-Protection": "1; mode=block"
                    },
                    "status_code": 200,
                    "title": " | "
                },
                "tcp_connect": {
                    "163.44.187.153:443": {
                        "failure": null,
                        "status": true
                    }
                }
            },
            "control_failure": null,
            "dns_consistency": "consistent",
            "dns_experiment_failure": null,
            "headers_match": true,
            "http_experiment_failure": null,
            "queries": [
                {
                    "answers": [
                        {
                            "answer_type": "CNAME",
                            "hostname": "www.iconnecthere.com",
                            "ttl": null
                        },
                        {
                            "answer_type": "A",
                            "ipv4": "163.44.187.153",
                            "ttl": null
                        }
                    ],
                    "engine": "system",
                    "failure": null,
                    "hostname": "www.iconnecthere.com",
                    "query_type": "A",
                    "resolver_hostname": null,
                    "resolver_port": null
                }
            ],
            "requests": [
                {
                    "failure": null,
                    "request": {
                        "body": "",
                        "headers": {
                            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
                            "Accept-Language": "en-US;q=0.8,en;q=0.5",
                            "User-Agent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36"
                        },
                        "method": "GET",
                        "tor": {
                            "exit_ip": null,
                            "exit_name": null,
                            "is_tor": false
                        },
                        "url": "https://www.iconnecthere.com/"
                    },
                    "response": {
                        "body": "...",
                        "code": 200,
                        "headers": {
                            "Connection": "keep-alive",
                            "Content-Type": "text/html; charset=UTF-8",
                            "Date": "Sun, 16 Jun 2019 07:54:49 GMT",
                            "Link": "<https://www.iconnecthere.com/wp-json/>; rel=\"https://api.w.org/\", <https://wp.me/P9whHu-Bz>; rel=shortlink",
                            "Server": "nginx",
                            "Transfer-Encoding": "chunked",
                            "X-Content-Type-Options": "nosniff",
                            "X-Nginx-Cache": "HIT",
                            "X-XSS-Protection": "1; mode=block"
                        },
                        "response_line": "HTTP/1.1 200 OK"
                    }
                }
            ],
            "retries": null,
            "socksproxy": null,
            "status_code_match": true,
            "tcp_connect": [
                {
                    "ip": "163.44.187.153",
                    "port": 443,
                    "status": {
                        "blocked": false,
                        "failure": null,
                        "success": true
                    }
                }
            ],
            "title_match": false
        },
        "test_name": "web_connectivity",
        "test_runtime": 7.021033048629761,
        "test_start_time": "2019-06-16 07:54:47",
        "test_version": "0.0.1"
    },
    "format": "json"
}
[engine] jsonapi: method: POST
[engine] jsonapi: URL: https://c.collector.ooni.io:443/report/20190616T075447Z_AS30722_uYWBiL1DmELYYLTFKUmixgZhNwfCQxHm6Nlhj5Q4Wk8DMYJ3Ir
[engine] http: connection to 37.218.242.210:443 ready; sending request
```

Note that we're again reusing the open connection.

```
[engine] > POST /report/20190616T075447Z_AS30722_uYWBiL1DmELYYLTFKUmixgZhNwfCQxHm6Nlhj5Q4Wk8DMYJ3Ir HTTP/1.1
[engine] > Content-Length: 595830
[engine] > Host: c.collector.ooni.io:443
[engine] > Content-Type: application/json
[engine] > User-Agent: ooniprobe-desktop/3.0.0-beta.3
[engine] > Accept-Encoding: gzip
[engine] >
[engine] http: request sent; waiting for response
[engine] http: start receiving response
[engine] < HTTP/1.1 200 OK
[engine] < Date: Sun, 16 Jun 2019 07:54:55 GMT
[engine] < Content-Type: application/json
[engine] < Content-Length: 21
[engine] < Connection: keep-alive
[engine] < Keep-Alive: timeout=120
[engine] < Server: nginx/1.6.2
[engine] <
[engine] jsonapi: response body: {"status": "success"}
```

And this is it for the first URL, now we're going over the second URL:

```
   • Fetching: 0 &{
    32 web_connectivity 2019-06-16 07:54:47.781657 +0000 UTC
    7.814443 true true false { false} false { false} false {... true} {198 true} {0 false} {
    false false}  13 /.../ooni283065313/msmt-websites.WebConnectivity-2019-06-16T075447.134412Z.jsonl}
   • status.measurement_start 
[engine] measurementkit: settings: {
    "inputs":["http://adultfriendfinder.com/"],
    "log_level":"DEBUG",
    "name":"WebConnectivity",
    "options":{
...
}}
```

So, it should be clear now that we're not performing a geolookup for every URL and we're not opening a report for every URL. Rather, these functionality have been moved from the C++ engine to the Go engine. This is part of the plan to move as much functionality as possible to the Go engine, to make the C++ engine unnecessary in the future.

Do you agree with this analysis? If so, I believe the correct follow-up is to address all the issues that I've found while reading the logs before merging.

Reply:
Yes what you say makes sense and apologies for jumping to conclusions. I took note of this last night mostly so I wouldn't forget to look into with more calm.

Thank you for your very detailed and thorough analysis which seems very reasonable.

What tripped me off was that I was seeing in the logs:
```
[engine] Your public IP address: 188.152.18.76
[engine] Your country: IT
[engine] Your ISP identifier: AS30722
[engine] Your ISP name: Vodafone Italia S.p.A.
   • OnProgress: 0.050000 - geoip lookup
   7.05% geoip lookup
```

Which made me believe that a geoip lookup was being done each time (this log line did not appear in previous versions). Perhaps we should remove the line saying "geoip lookup" from the logs in this case.

Similarly for report creation I was tipped off by the log line saying:
```
   6.10% open report
```

on each URL.

Perhaps we can cleanup the output a bit to show these messages only when the report is opened for the first time and the geoip lookup operation is being performed?

Reply:
Indeed! In fact, I removed these logs from MK v0.10.4 precisely because I thought they were rather confusing :-).

Reply:
@hellais:

> Edit: I should also add that I noticed that the runtime of the run websites command on this branch appeared to be slower than on the current master.

The first explanation that comes to mind is that MK uses parallelism `3` whereas the code I've added in here is strictly sequential. As you know, we've discussed a number of times what to do in terms of parallelism, and it seems on Monday with @FedericoCeratto and you we concluded that we're fine with parallelism `1` by default. Hence, this initial iteration of the code has no provision for parallelism and all is sequential.

Reply:
I have made sure we have actionable issues for every aspect mentioned in this branch of the conversation, so I am going to resolve it, to simplify my understanding of progress.